pandas/core/dtypes/concat.py

```diff
@@ -108,7 +173,17 @@ def is_nonempty(x) -> bool:
         # we ignore axis here, as internally concatting with EAs is always
         # for axis=0
         if not single_dtype:
-            target_dtype = find_common_type([x.dtype for x in to_concat])
+            # Special case for handling concat with categorical series.
+            # We need to make sure that categorical dtype is preserved
+            # when an array of valid values is given (GH#25383)
+            if (
+                isinstance(to_concat[0], ExtensionArray)
+                and all(x.shape[0] == 1 for x in to_concat[1:])
+                and _can_cast_to_categorical(to_concat)
```

Comment:
this a very complicated implementation. This should *all* be in `find_common_type`` , but should be *much* simpler that this. either the dtypes are the same or they are not. changing them is not in scope for this issue.

Reply:
Ok, scope is more clear to me now. I will revert back to the previous approach and adapt it to raise on `dtype` mismatch.